### PR TITLE
nix: bump to bazel 7.1

### PR DESCRIPTION
**Pull request by Strum355**
Using my fork with open PR: https://github.com/NixOS/nixpkgs/pull/295615

Im needing 7.1 sooner than later, given its new features and this repo also using 7.1 (so all new features are fair game for it!)

Cached version is pushed to sourcegraph-noah.cachix.org, which I've added at the top here so you too can avoid building it locally :slightly_smiling_face: 

## Test plan

Tested locally :sunglasses:

### CODE REVIEW

**Review by Chickensoupwithrice — APPROVED** [review]
Able to build locally, and run `sg start`

**Comment by Strum355** [comment]
> Able to build locally, and run `sg start`

Fwiw we dont use nixpkgs' bazel on macos, so this only affects linux 😅

File: flake.nix [hunk 1]
```diff
@@ -1,14 +1,20 @@
 {
   description = "The Sourcegraph developer environment & packages Nix Flake";
 
+  nixConfig = {
+    extra-substituters = [ "https://sourcegraph-noah.cachix.org" ];
+    extra-trusted-public-keys = [ "sourcegraph-noah.cachix.org-1:rTTKnyuUmJuGt/UAXUpdOCOXDAfaO1AYy+/jSre3XgA=" ];
+  };
```
Comment (keegancsmith) [inline flake.nix:7]: <3

Comment (keegancsmith) [reply flake.nix:7]: On my amd64 linux machine it still seemed to build bazel-7.1.0

Comment (Strum355) [reply flake.nix:7]: Did it prompt you to trust this substituter? Can you figure out what the store path is and I can check if it matches what Id expect? `nix path-info .#bazel_7` I think

Comment (keegancsmith) [reply flake.nix:7]: right, I answered yes but just noticed this warning: 

```
warning: ignoring untrusted substituter 'https://sourcegraph-noah.cachix.org', you are not a trusted user.
```

```
❯ nix path-info .#bazel_7
warning: ignoring untrusted substituter 'https://sourcegraph-noah.cachix.org', you are not a trusted user.
Run `man nix.conf` for more information on the `substituters` configuration option.
evaluating derivation 'git+file:///home/keegan/src/github.com/sourcegraph/sourcegraph#bazel_7'/nix/store/n63v6d0m4p32n7zl3aqvmdnzq3m0ajm4-bazel-7.1.0
```

Comment (Strum355) [reply flake.nix:7]: That path exists, so its a case of you needing to add your user to the list of trusted users in nix.conf, then itll work

Comment (keegancsmith) [reply flake.nix:7]: Fixed the warning without making my user effectively root :) https://github.com/keegancsmith/dotfiles/commit/8936f4415ed4a6b46a1569da4f966e629cee1dc7

Comment (Strum355) [reply flake.nix:7]: Just finished reading about that as well, wew it is confusing!

Comment (Strum355) [reply flake.nix:7]: Annoyingly, the CLI prompts for adding the substituter as trusted if it isnt already, but only if youre a trusted user I presume, which makes the experience a bit better than you having to edit nix.conf :/

**Pull request merged**